core/vm/interpreter.go

```diff
@@ -228,6 +262,16 @@ func (in *EVMInterpreter) Run(contract *Contract, input []byte, readOnly bool) (
 			in.cfg.Tracer.CaptureState(pc, op, gasCopy, cost, callContext, in.returnData, in.evm.depth, err)
 			logged = true
 		}
+		// fail if op is not available in legacy context
+		if operation.eof1 && callContext.Contract.IsLegacy() {
+			_, err = opUndefined(&pc, in, callContext)
+			break
+		}
+		// fail if op is not available in eof context
+		if operation.legacyOnly && !callContext.Contract.IsLegacy() {
+			_, err = opUndefined(&pc, in, callContext)
+			break
```

Comment:
This is an interesting design choice -- having a shanghai jumptable, where each op has two booleans: 

- `eof1` true iff the op is exclusive for eof-contracts, 
- `legacyOnly` true iff the op is exclusive to non-eof contracts. 

I think the naming is a bit inconsistent? `onlyEof/notEof`, or `onlyEof/onlyLegacy`, or `eofOnly/legacyOnly`. 

Anyway, another way to do it would be to have the dimensionality at another level, so we use one jumptable for eof-contracts, and another jumptable for non-eof contracts. Then we'd have 

```
shanghaiEofInstructionSet   = newShanghaiInstructionSet( true) 
shanghaiLegacyInstructionSet   = newShanghaiInstructionSet( false ) 
```
It might be the case that the `shanghaiLegacyJumptable` is actually the good ole `londonInstructionSet`, if nothing legacy changes (well, currently `TSTORE`/`TLOAD` would make it into both instructionsets). 

Do you see any particular downsides to doing it on the jumptable-level instead of the `opCode` level?

Reply:
Agreed on moving to jump table. I just wanted to have it here for now because it is very simple and allows me to focus on actual changes for EOF. I'll definitely refactor the jump table later for this. Not sure if that makes sense in a separate PR or just in this one. Maybe it's a small change, I'll see if I can sketch something out.